res/controllers/Traktor.Kontrol-Z2-hid-scripts.js

```diff
@@ -0,0 +1,629 @@
+    engine.softTakeover("[EqualizerRack1_[Channel1]_Effect1]", "parameter2", true);
+    engine.softTakeover("[EqualizerRack1_[Channel1]_Effect1]", "parameter3", true);
+    engine.softTakeover("[EqualizerRack1_[Channel2]_Effect1]", "parameter1", true);
+    engine.softTakeover("[EqualizerRack1_[Channel2]_Effect1]", "parameter2", true);
+    engine.softTakeover("[EqualizerRack1_[Channel2]_Effect1]", "parameter3", true);
+   
+    // engine.softTakeover("[Master]", "crossfader", true);
+    engine.softTakeover("[Master]", "gain", true);
+    engine.softTakeover("[Master]", "headMix", true);
+    engine.softTakeover("[Master]", "headGain", true);
+
+    // Dirty hack to set initial values in the packet parser
+    var data = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+    TraktorZ2.incomingData(data);
+};
+
+TraktorZ2.registerInputScaler = function(message, group, name, offset, bitmask, callback) {
+	HIDDebug("TraktorZ2: registerInputScaler");
+    message.addControl(group, name, offset, "H", bitmask);
+    message.setCallback(group, name, callback);
+};
+
+TraktorZ2.registerInputButton = function(message, group, name, offset, bitmask, callback) {
+	HIDDebug("TraktorZ2: registerInputButton");
+    message.addControl(group, name, offset, "B", bitmask);
+    message.setCallback(group, name, callback);
+};
+
+TraktorZ2.shiftHandler = function(field) {
+    HIDDebug("TraktorZ2: shiftHandler");
+	engine.setValue("[Controls]", "touch_shift", field.value);
+    this.shiftPressed = field.value;
+   // TraktorZ2.basicOutputHandler(field.value, field.group, "!shift");
+};
+
+TraktorZ2.parameterHandler = function(field) {
+	HIDDebug("TraktorZ2: parameterHandler");
+    engine.setParameter(field.group, field.name, field.value / 4095);
+};
+
+// TraktorZ2.filterHandler = function(field) {
+	// HIDDebug("TraktorZ2: filterHandler");
+    // // The super knob drives all the supers!
+    // var chan = TraktorZ2.Channels[field.group];
+    // var value = field.value / 4095.;
+    // // if (chan.fxEnabledState) {
+        // // for (var fxNumber = 1; fxNumber <= 4; fxNumber++) {
+            // // if (TraktorZ2.fxButtonState[fxNumber]) {
+                // // engine.setParameter("[EffectRack1_EffectUnit" + fxNumber + "]", "super1", value);
+            // // }
+        // // }
+    // // } else {
+        // engine.setParameter("[QuickEffectRack1_" + chan.group + "]", "super1", value);
+    // //}
+// };
+
+TraktorZ2.messageCallback = function(_packet, data) {
+    for (var name in data) {
+        if (Object.prototype.hasOwnProperty.call(data, name)) {
+            TraktorZ2.controller.processButton(data[name]);
+        }
+    }
+};
+
+TraktorZ2.incomingData = function(data, length) {
+    TraktorZ2.controller.parsePacket(data, length);
+};
+
+TraktorZ2.shutdown = function() {
+    // Deactivate all LEDs
+    var dataStrings = [
+        "      00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +

```

Comment:
there is a bug in this code (that was in my code too-- if you have leading spaces, the first item in the list is an empty string so that byte does get set

Reply:
the fix is to eliminate leading spaces in your strings

Reply:
Thank's for poiting to this! This is the shutdown code, which I've to rework anyway. Currently I'm working on debugLights.

The shutdown code for the Z2 will be tricky anyway, because of the concept of seamless switching between internal and external mixing mode. The idea is:
-At the end of your DJ set, switch one channel from internal to external mixing mode and play an analog vinyl
-DJ1 disconnects his laptop -> Shutdown code must not interfere the music
-DJ2 connects his laptop -> Init code must not interfere the music
-Switch the deck with the analog vinyl back to internal mixing mode

Therefore I will have to decide LED by LED if the shutdown code should reset it, or if the LED represents a still active state.